<a href="https://colab.research.google.com/github/SzymonNowakowski/ml_workshop_jupyter/blob/main/0_tensors.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Let's talk about Tensors

A PyTorch or TensorFlow Tensor is just a **multidimensional array**. But it is much more transformation-centered, and when I say *transformation* I mean mathematical transformation.

Many times in the past it proved for me hard to speak clearly of Tensor transformations and dimensions. Below is a screenshot from a random tutorial, which shows a clear confusion when the author talks about Tensor dimensions.

![Tensor dimensions confusion - screenshot from https://beerensahu.wordpress.com/2018/03/21/pytorch-tutorial-lesson-1-tensor/](https://i.imgur.com/jikCq6K.png)

In [ ]:
torch.Tensor(2,3)

tensor([[-1.1022e+37,  4.5573e-41, -1.1022e+37],
        [ 4.5573e-41,  0.0000e+00,  0.0000e+00]])

Considering that a Tensor is a much more mathematically inclined object than a multidimensional array, let me introduce a naming convention which will be compatible with mathematical objects a Tensor represents. For instance, a vector from $\mathbb{R}^n$ is considered $n$-dimensional in mathematics but it can be stored in a one-dimensional array, a matrix in $\mathbb{R}^{r \times c}$ is considered to be $r \times c$-dimensional  object, but can be stored in a two-dimensional array. This duality creates a confusion. In my experience the following naming convention is coherent considering that Tensors **are** mathematical objects.

Tensor Terms  | Meaning | Multidimensional Array Terms
---|---|---
**order** | number of indices (levels) in a Tensor | dimension
**dimension** | number of components a Tensor can store in a given order | size
Tensor of order zero | a constant | - |
Tensor of order one | a vector | one-dimensional array |
Tensor of order two | a matrix | two-dimensional array |
`torch.tensor([1.1, 4.12, 8.9, 14.85])` | an order-one 4-dimensional tensor representing a vector in $\mathbb{R}^4$ | `[1.1, 4.12, 8.9, 14.85]`